Discrepancy between Github and Travis test results / Warnings count during compilation

On Travis, this minimal build failed

https://travis-ci.org/github/nest/nest-simulator/jobs/761653390

while it passed on Github

https://github.com/nest/nest-simulator/pull/1966/checks?check_run_id=2044070265

Also the other tests (MPI_ONLY, OPENMP_ONLY, FULL) fail on Travis but pass on Github.

The reason for the failure on Travis is that the warnings count for `sli/scanner.cc` is 3 instead of the expected 0. The compiler emits warnings on Travis and Github, but for some reason they are not registered properly by our result handler.

The test failed on Travis because changes in line numbers in `scanner.cc` lead to `known_warnings` no longer being recognized. I fixed that with an update to #1966. The reason that the Github checks did not pick up the problem is probably that the warnings regex define in https://github.com/nest/nest-simulator/blob/fc645dd297fac6b75d866de4802f4dec5bf7b65a/extras/parse_travis_log.py#L370 does not match on Github. The regex is
```python
nest_warning_re = re.compile(f'{build_dir}.*: warning:')
```
and the line to match in the build log
```
2021-03-06T00:44:11.9495242Z /home/runner/work/nest-simulator/nest-simulator/sli/scanner.cc:642:13: warning: this statement may fall through [-Wimplicit-fallthrough=]
```
So the problem is the timestamp which Github adds to each log line (click the three dots button in the top right corner of the fancy build log to get to the raw version with timestamps).

We'd need to adjust the regex and than also adjust the mechanism for matching the known warnings, so that will not be thrown off by the timestamps at the beginning of the lines.

Reassigning to @jougs and @gtrensch who worked last on this.

--- a/extras/parse_travis_log.py
+++ b/extras/parse_travis_log.py
@@ -393,7 +393,7 @@ def makebuild_summary(log_filename, msg_make_section_start,
                 # Only count warnings originating in NEST source files
                 warning_match = nest_warning_re.match(line)
                 if warning_match is not None:
-                    warning = warning_match.group(0)
+                    warning = warning_match.group(1)
                     if warning not in known_warnings:
                         file_name = warning.split(':')[0]
                         if file_name not in warning_summary:

--- a/extras/parse_travis_log.py
+++ b/extras/parse_travis_log.py
@@ -367,7 +367,7 @@ def makebuild_summary(log_filename, msg_make_section_start,
     # with some warnings, this would be a good point to re-set the
     # expected_warnings variable conditionally for that build_type.
 
-    nest_warning_re = re.compile(f'{build_dir}.*: warning:')
+    nest_warning_re = re.compile(f'.* ({build_dir}.*: warning:.*)')
     known_warnings = [
         f'{build_dir}/sli/scanner.cc:642:13: warning: this statement may fall through [-Wimplicit-fallthrough=]',
         f'{build_dir}/sli/scanner.cc:673:19: warning: this statement may fall through [-Wimplicit-fallthrough=]',
@@ -392,12 +392,14 @@ def makebuild_summary(log_filename, msg_make_section_start,
 
                 # Only count warnings originating in NEST source files
                 warning_match = nest_warning_re.match(line)
-                if warning_match is not None and line.strip() not in known_warnings:
-                    file_name = line.split(':')[0]
-                    if file_name not in warning_summary:
-                        warning_summary[file_name] = 0
-                    warning_summary[file_name] += 1
-                    number_of_warning_msgs += 1
+                if warning_match is not None:
+                    warning = warning_match.group(0)
+                    if warning not in known_warnings:
+                        file_name = warning.split(':')[0]
+                        if file_name not in warning_summary:
+                            warning_summary[file_name] = 0
+                        warning_summary[file_name] += 1
+                        number_of_warning_msgs += 1
 
                 if is_message(line, msg_make_section_end):
                     # The log file contains only one 'make' section, return.

